cudamapper/src/index_gpu_two_indices.cu

```diff
@@ -0,0 +1,129 @@
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "index_gpu_two_indices.cuh"
+
+namespace claragenomics
+{
+namespace cudamapper
+{
+namespace details
+{
+namespace index_gpu_two_indices
+{
+void find_first_occurrences_of_representations(thrust::device_vector<representation_t>& unique_representations_d,
+                                               thrust::device_vector<std::uint32_t>& first_occurrence_index_d,
+                                               const thrust::device_vector<representation_t>& input_representations_d)
+{
+    // each element has value 1 if representation with the same index in representations_d has a different value than it's neighbour to the left, 0 otehrwise
+    // underlying type is 32-bit because a scan operation will be performed on the array, so the elements should be capable of holding a number that is equal to
+    // the total number of 1s in the array
+    thrust::device_vector<std::uint32_t> new_value_mask_d(input_representations_d.size());
+
+    // TODO: Currently maximum number of thread blocks is 2^31-1. This means we support representations of up to (2^31-1) * number_of_threads
+    // With 256 that's (2^31-1)*2^8 ~= 2^39. If representation is 4-byte (we expect it to be 4 or 8) that's 2^39*2^2 = 2^41 = 2TB. We don't expect to hit this limit any time soon
+    // The kernel can be modified to process several representation per thread to support arbitrary size
+    std::uint32_t number_of_threads = 256; // arbitrary value
+    std::uint32_t number_of_blocks  = (input_representations_d.size() - 1) / number_of_threads + 1;
+
+    create_new_value_mask<<<number_of_blocks, number_of_threads>>>(input_representations_d.data().get(),
+                                                                   input_representations_d.size(),
+                                                                   new_value_mask_d.data().get());
+    CGA_CU_CHECK_ERR(cudaDeviceSynchronize()); // sync not necessary, here only to detect the error immediately
+
+    // do inclusive scan
+    // for example for
+    // 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20
+    // 0  0  0  0 12 12 12 12 12 12 23 23 23 32 32 32 32 32 46 46 46
+    // 1  0  0  0  1  0  0  0  0  0  1  0  0  1  0  0  0  0  1  0  0
+    // gives
+    // 1  1  1  1  2  2  2  2  2  2  3  3  3  4  4  4  4  4  5  5  5
+    // meaning all elements with the same representation have the same value and those values are sorted in increasing order starting from 1
+    thrust::device_vector<std::uint64_t> representation_index_mask_d(new_value_mask_d.size());
+    thrust::inclusive_scan(thrust::device,
```

Comment:
I have the feeling the `create_new_value_mask` and this `inclusive_scan` could be combined into a `transformed_inclusive_scan` on the index of `input_representations_d` (via `thrust::counting_iterator`). This could save us the whole `new_value_mask_d` data structure including all related memory transfers.

Reply:
This might be postponed to a separate PR.

Reply:
That could be the case, but let's leave it for a separate PR